sessions/forecasting-nowcasting.qmd

```diff
@@ -0,0 +1,798 @@
+  By jointly modelling the entire process, it can provide more coherent and potentially more accurate forecasts, especially when the reporting process is well understood.
+
+- **Why might it perform worse than the complete data approach?**  
+  In some situations, the complete data approach, which waits for older data to become nearly complete before forecasting, might perform better.  
+  This can happen if the most recent data are highly incomplete or unrepresentative, for example, due to sudden changes in reporting or the epidemic process.  
+  In such cases, older data may provide a more stable basis for forecasting, even though it is less timely.  
+  It may also underperform if the generative process is not consistent with the data we have. For example, if the growth rate of older data happens to be more consistent with the future growth rate of the epidemic, then the complete data approach will perform better.
+  These issues indicate that the underlying generative process or forecast model could be improved to better capture changes in the epidemic or reporting dynamics.
+
+:::
+
+# Evaluation and comparison
+
+We'll evaluate all four approaches against what will finally be reported (our `complete_at_horizon` dataset) using proper scoring rules.
+
+## Prepare evaluation data
+
+```{r prepare-evaluation}
+# Create evaluation dataset - what we're forecasting against
+evaluation_data <- complete_at_horizon |>
+  filter(day > cutoff, day <= cutoff + forecast_horizon) |>
+  mutate(
+    forecast_horizon = day - cutoff,
+    observed = complete_onsets
+  ) |>
+  select(day, forecast_horizon, observed)
+
+head(evaluation_data)
+```
+
+## Visualise all the forecasts together
+
+As in all other sessions, we first visualise the forecasts together.
+
+```{r plot-all-forecasts}
+forecast_results <- bind_rows(
+  complete_data_forecasts |>
+    filter(actual_day > cutoff),
+  pipeline_forecasts,
+  joint_forecasts
+)
+
+forecast_results |>
+  ggplot(aes(x = actual_day, y = .value, group = .draw, color = approach)) +
+  geom_line(alpha = 0.1) +
+  geom_point(data = evaluation_data, aes(x = day, y = observed, group = NULL, color = NULL), color = "black", size = 1) +
+  scale_y_log10() +
+  labs(x = "Day", y = "Onsets", title = "All forecast approaches") +
+  theme_minimal() +
+  guides(color = guide_legend(
+    title = "Approach", override.aes = list(alpha = 1, linewidth = 1))
+  ) +
+  theme(legend.position = "bottom")
+```
+
+## Quantitative evaluation using scoringutils
+
+Using `scoringutils`, We first create a forecast sample object
+
+```{r forecast-evaluation}
+sc_forecasts <- forecast_results |>
+    left_join(evaluation_data, by = c("actual_day" = "day")) |>
+    group_by(approach, actual_day) |>
+    mutate(
+      .draw = row_number()
+    ) |>
+    as_forecast_sample(
+      forecast_unit = c("approach", "actual_day"),
+      observed = "observed",
+      predicted = ".value",
+      sample_id = ".draw"
+    )
+sc_forecasts
+```
```

Comment:
``` 
Warning: Some forecasts have different numbers of rows (e.g. quantiles or samples).
scoringutils found: 98, 80. This may be a problem (it can potentially distort
scores, making it more difficult to compare them), so make sure this is
intended.
```

Does this mean that some models had more forecast horizons than others? if so, maybe consider aligning them so it's the same?